### DISCUSSION TITLE
[@types/react] useEffect types are not strict enough to prevent page crashes

### DISCUSSION THREAD

**Discussion by noahtallen** [DefinitelyTyped/DefinitelyTyped#65890] [Issues with a @types package]
Hi! I came across an issue today where an incorrect `useEffect` return value caused a production crash.

The code looked similar to this (though `ifCheck` and `doSomething` are contrived):

```ts
// This value comes from something that isn't typed very strongly like an old redux store
const ifCheck = someValueInferredAsAny;
const doSomething = () => 'string';

useEffect( () => ifCheck && doSomething(), [ ... ] );
```

This passes type checks, because the callback is inferred as `() => any`, even though there are cases where it will return "string".

I'm not sure if there's a way to clarify the types here, but it would be very useful. It's easy to solve this by using `() => { ifCheck && doSomething() }` instead, but there is no automated way to find these issues in large, older codebases without stronger types or a lint rule.

This issue occurred after updating to React 18, and the code wasn't crashing before that. While there are browser warnings now, they must not have been caught during development, before we were on React 18. Automated CI checks are a much better way to get people to notice and fix these types of issues, which is why stronger types would be super helpful for us!

[a comment by an automated account is not shown]

**Comment by eps1lon** [comment]
The types cannot solve this for this and neither can TypeScript since `ifCheck` is `any`. If you make sure it's `unknown`, [it doesn't type-check as expected](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbwK4GcCmBRAZltBjGAXziyghDgHIo0BDAygKEYBN8AbWmuPCAOxTxgWAMIALfAGsAXHCR9JfCAHc+Abka8B8FhADK5NDDHA+AczgBeOAAoAlFYB8VQVFNnKGxqkw58MG1sHS2dhcSk4ADJIuF0DECMTc3s7NSA). Remember, if `ifCheck` is a non-empty string you get crashes as well.
